src/cscapi/client.py

```diff
@@ -0,0 +1,276 @@
+                except httpx.HTTPStatusError as exc:
+                    logging.error(
+                        f"Error while sending metrics: {exc} for machine {machine_to_process.machine_id}"
+                    )
+
+            attempt_count += 1
+            machines_to_process_attempts = next_machines_to_process_attempts
+
+    def _send_signals(self, token: str, signals: SignalModel):
+        for signal_batch in batched(signals, 250):
+            body = [asdict(signal) for signal in signal_batch]
+            resp = self.http_client.post(
+                CAPI_SIGNALS_URL, json=body, headers={"Authorization": token}
+            )
+            resp.raise_for_status()
+
+    def _send_metrics_for_machine(self, machine: MachineModel):
+        resp = self.http_client.post(
+            CAPI_METRICS_URL,
+            json={
+                "bouncers": [],
+                "machines": [
+                    {
+                        "last_update": datetime.datetime.now().isoformat(),
+                        "last_push": datetime.datetime.now().isoformat(),
+                        "version": __version__,
+                        "name": machine.machine_id,
+                    }
+                ],
+            },
+        )
+        resp.raise_for_status()
+
+    def _prune_signals(self):
+        signals = filter(lambda signal: signal.sent, self.storage.get_all_signals())
+        self.storage.delete_signals(signals)
+
+    def _refresh_machine_token(self, machine: MachineModel) -> MachineModel:
+        resp = self.http_client.post(
+            CAPI_WATCHER_LOGIN_URL,
+            json={
+                "machine_id": machine.machine_id,
+                "password": machine.password,
+                "scenarios": machine.scenarios.split(","),
+            },
+        )
+        try:
+            resp.raise_for_status()
+        except httpx.HTTPStatusError as exc:
+            logging.error(
+                f"Error while refreshing token: machine_id might be already registered or password is wrong"
+            )
+            raise exc
+
+        new_machine = asdict(machine)
+        new_machine["token"] = resp.json()["token"]
+        new_machine = MachineModel(**new_machine)
+        self.storage.update_or_create_machine(new_machine)
+        return new_machine
+
+    def _register_machine(self, machine: MachineModel) -> MachineModel:
+        logging.info(f"registering machine {machine.machine_id}")
+        machine.password = (
+            machine.password if machine.password else secrets.token_urlsafe(32)
+        )
+        resp = self.http_client.post(
+            CAPI_WATCHER_REGISTER_URL,
+            json={
+                "machine_id": machine.machine_id,
+                "password": machine.password,
+            },
+        )
+        self.storage.update_or_create_machine(machine)
+        return machine
+
+    def _make_machine(self, machine: MachineModel):
```

Comment:
Function starting with _ are internal functions and not meant for user to call. That being said it's better to have good naming. 

I don't think splitting is good idea since the old version was kinda like that and you asked to factorize it. We anyways need to do both at the same time anyways.

Reply:
If i remember correctly, the old version had the code duplicated in send signals and get decisions. That's not what I'm suggesting here.
also it seems my comment was unfinished.

the make machine would only create of retrieve non failing machines
The ensure_machine_is_logged would check and retrieve the token

the ensure function can be called at the beginning of the loop for dealing with a specific machine signals (this way the token is fresh)

Reply:
I mean having the ensure function right after checking it's not failing in for 

"machine_to_process in machines_to_process_attempts:"